tests/lang/s02_core/s02_procedures/t03_overload_core.nim

```diff
@@ -0,0 +1,210 @@
+  ## Subtype match - argument is a subtype of the procedure argument. Note that this
+  ## requirement is related to inheritance (because of subtyping), but does not
+  ## replace `method`'s functionality. It also works with `{.inheritable.}` types and
+  ## `object of RootObj` (in addition to `ref object of RootObj`) whereas `method`
+  ## requires `ref` to be used.
+  block inheritable_pragma:
+    ## `{.inheritable.}` provides subtyping capabilities
+    type
+      Base {.inheritable.} = object
+        field: int
+
+      Derived = object of Base
+        field2: int
+
+    proc impl(arg: Base) = discard
+
+    impl(Derived())
+    impl(Base())
+
+    ## Note that only embedded supertype part is passed to the procedure in this
+    ## case, and it is not safe to `cast` to a base type.
+    ##
+    ## ```
+    ## Derived[Sup = Base[field1: int], field2: int]
+    ##         ^^^^^^^^^^^^^^^^^^^^^^^
+    ##         Only this part of the object is passed to the procedure.
+    ## ```
+
+  block subtype_ref_ptr_var:
+    ## `ref` and `ptr` and `var` do not affect subtype relation - as a result they
+    ## can also be used to call a procedure that expects a supertype.
+    type
+      Base = object of RootObj
+        fbase: int
+
+      Derived = object of Base
+        fderived: int
+
+    proc impl(arg: ptr Base): string = "ptr"
+    proc impl(arg: ref Base): string = "ref"
+    proc impl(arg: var Base): string = arg = Base(fbase: 256); "var"
+
+    doAssert impl((ptr Derived)(nil)) == "ptr"
+    doAssert impl((ref Derived)(nil)) == "ref"
+
+    var derived = Derived(fderived: 128)
+
+    ## It is possible to modify part of the embedded supertype via `var` argument
+    doAssert impl(derived) == "var"
+    doAssert derived.fbase == 256, "Assigned only to a section of a type"
+    doAssert derived.fderived == 128, "Original field value is intact"
+
+
+  block closest_subtype:
+    type
+      A = ref object of RootObj
+      B = ref object of A
+      C = ref object of B
+
+    proc impl(b: B): string = "matched subtype B"
+    proc impl(a: A): string = "matched subtype A"
+
+    ## Closest subtype takes precedence
+    doAssert impl(C()) == "matched subtype B"
+    doAssert impl(B()) == "matched subtype B"
+
+
+
+
+block most_specific_generic_wins:
+  ## If multiple overloaded procedures are present, most specific is
+  ## always selected.
+
+  block qualifier:
+    ## `ref`, `ptr` and other type qualifiers are counted as a part of
+    ## the generic procedure definition.
+
+    proc gen[T](x: var ref ref T): string = "var ref ref T"
+    proc gen[T](x: ref ref T): string = "ref ref T"
+    proc gen[T](x: ref T): string = "ref T"
+    proc gen[T](x: T): string = "T"
+
+    var v: ref ref int = nil
+    doAssert gen(v) == "var ref ref T"
+    doAssert gen((ref ref int)(nil)) == "ref ref T"
+    doAssert gen((ref int)(nil)) == "ref T"
+    doAssert gen(nil) == "T"
+
+
+  # QUESTION - it is not possible to select less generic overload?
```

Comment:
right now you'd have to do that by removing type information, as in a `cast`.

this makes sense from a static polymorphic dispatch perspective, IMO.